x-pack/dev-tools/simulate_autoscaling.py

```diff
@@ -0,0 +1,276 @@
+from collections import deque, defaultdict
+import enum
+import heapq
+import math
+import random
+import matplotlib.pyplot as plt
+
+
+START_TIME = 0
+END_TIME = 5*3600
+MEASUREMENT_INTERVAL = 1
+
+AUTO_SCALE_UP_THRESHOLD = 0.9
+AUTO_SCALE_DOWN_THRESHOLD = 0.8
+
+AVG_INFERENCE_TIME = 0.1
+PHYSICAL_CORES_PER_NODE = 4
+
+
+random.seed(170681)
+
+
+class Estimator:
+  def __init__(self, value, variance, smoothing_factor):
+    self.value = value
+    self.variance = variance
+    self.smoothing_factor = smoothing_factor
+
+  def add(self, value, variance, dynamics_changed=False) -> None:
+    process_variance = variance / self.smoothing_factor
+
+    if dynamics_changed or abs(value - self.value)**2 / (self.variance + variance) > 100:
+      process_variance *= self.smoothing_factor
+      print(f'dynamic changed: value={value} state={self.value} +-/ {math.sqrt(self.variance)}')
+
+    gain = (self.variance + process_variance) / (self.variance + process_variance + variance)
+    self.value += gain * (value - self.value)
+    self.variance = (1 - gain) * (self.variance + process_variance)
+
+  def estimate(self) -> float:
+    return self.value
+
+  def lower(self):
+    return max(0.0, self.value - math.sqrt(self.variance))
+
+  def upper(self):
+    return self.value + math.sqrt(self.variance)
+
+
+class Simulator:
```

Comment:
As above I would give some motivation for this:
```suggestion
class Simulator:
    """
    We simulate a Poisson process for inference arrivals with time varying rate parameter. This models a
    variety of rate dynamics: smooth ramp, smooth periodic, shock and steady.

    Inference is modelled as mean duration + uniform noise. The mean duration captures the behaviour of
    vCPUs which is that throughput is largely constant after all physical cores on a node are occupied.
    
    We assume perfect load balancing, i.e. that every allocation of inference is never waiting whilst there are
    inferences to be done. Inferences can only be picked off once they have arrived and allocations can only
    pick an inference "inference duration" after they picked their last inference. We assume they are selected
    FIFO.
    
    The key user parameters are the number of waiting inferences and the average and maximum delay to
    receive each inference which can be calculated from the difference between when inference calls arrived
    and when they are available.
    """
```